go/codegen/experimental/xgboost.go

```diff
@@ -0,0 +1,327 @@
+// Copyright 2020 The SQLFlow Authors. All rights reserved.
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package experimental
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+	"text/template"
+
+	"sqlflow.org/sqlflow/go/attribute"
+	"sqlflow.org/sqlflow/go/codegen"
+	tf "sqlflow.org/sqlflow/go/codegen/tensorflow"
+	"sqlflow.org/sqlflow/go/ir"
+	pb "sqlflow.org/sqlflow/go/proto"
+)
+
+type xgbTrainFiller struct {
+	DataSource         string
+	Select             string
+	ValidationSelect   string
+	ModelParamsJSON    string
+	TrainParamsJSON    string
+	FieldDescJSON      string
+	LabelJSON          string
+	FeatureColumnNames []string
+	FeatureColumnCode  string
+	DiskCache          bool
+	BatchSize          int
+	Epoch              int
+	Submitter          string
+}
+
+// XGBoostGenerateTrain returns the step code
+func XGBoostGenerateTrain(trainStmt *ir.TrainStmt, session *pb.Session) (string, error) {
+	if err := resolveModelParams(trainStmt); err != nil {
+		return "", err
+	}
+	params := parseAttribute(trainStmt.Attributes)
+	diskCache := params["train."]["disk_cache"].(bool)
+	delete(params["train."], "disk_cache")
+
+	var batchSize, epoch = -1, 1
+	batchSizeAttr, ok := params["train."]["batch_size"]
+	if ok {
+		batchSize = batchSizeAttr.(int)
+		delete(params["train."], "batch_size")
+	}
+	epochAttr, ok := params["train."]["epoch"]
+	if ok {
+		epoch = epochAttr.(int)
+		delete(params["train."], "epoch")
+	}
+	if _, ok := params["train."]["num_workers"]; ok {
+		delete(params["train."], "num_workers")
+	}
+
+	// TODO(typhoonzero): use feature derivation at runtime.
+	if len(trainStmt.Features) != 1 {
+		return "", fmt.Errorf("xgboost only support 0 or 1 feature column set, received %d", len(trainStmt.Features))
+	}
+
+	featureColumnCode, featureFieldDesc, labelFieldDesc, err := deriveFeatureColumnCodeAndFieldDescs(trainStmt.Features["feature_columns"], trainStmt.Label)
```

Comment:
Will the feature derivation move to Python later?

Reply:
Yes, there are todo comments here and will be done in the next PR.